docs/TPPDialect.md

```diff
@@ -0,0 +1,351 @@
+  tpp.zero outs(%0) : memref<MxNxTy> // XOR op
+```
+
+## Copy
+Copies memory into a new place.
+Types must have the same shape.
+```mlir
+  %1 = memref.alloc : memref<MxNxTy>
+  tpp.copy ins(%0) outs(%1) : (memref<MxNxTy>, memref<MxNxTy>) // IDENTITY op
+  tpp.copy ins(%0) outs(%1) : (memref<NxTy>, memref<MxNxTy>) // Error! No automatic broadcast
+```
+
+If `%0` is unused after this (dead), we can always just elide the alloc and the copy and replace with `%0`.
+
+## Broadcast
+Broadcasts scalar/row/col into new memory.
+Types must not have the same shape. Out > in.
+Any `unary`/`binary`/`ternary` op that uses a broadcast flag should use this op on each operand needed.
+```mlir
+  %1 = memref.alloc : memref<MxNxTy>
+  tpp.broadcast ins(%0) outs(%1) : (memref<2x1xf32>, memref<2x4xf32>) // COL
+  tpp.broadcast ins(%0) outs(%1) : (memref<2xf32>, memref<4x2xf32>) // ROW
+  tpp.broadcast ins(%0) outs(%1) : (memref<f32>, memref<4x2xf32>) // SCALAR
+  tpp.broadcast ins(%0) outs(%1) : (f32, memref<4x2xf32>) // SCALAR
+  tpp.broadcast ins(%0) outs(%1) : (memref<2x4xf32>, memref<2x4xf32>) // Error, use COPY
+  tpp.broadcast ins(%0) outs(%1) : (memref<2x4xf32>, memref<2xf32>) // Error, use REDUCE
+```
+Should be fused with users.
+Should be safe to fuse with multiple users as they shouldn't alias output with an input of a different shape.
+It should be an error for `unary`/`binary`/`ternary` ops to alias with broadcast operands.
+Can be lowered as `COPY` with `BCAST` flags.
+
+## Reduce
+Reduce scalar/row/col into new memory.
+Types must not have the same shape. Out < in.
+Any `unary`/`binary`/`ternary` op that uses a reduce flag should use this op on each operand needed.
+```mlir
+  %1 = memref.alloc : memref<MxNxTy>
+  tpp.reduce_add ins(%0) outs(%1) : (memref<2x4xf32>, memref<2x1xf32>) // REDUCE_X_OP_ADD + REDUCE_COL
+  tpp.reduce_add ins(%0) outs(%1) : (memref<4x2xf32>, memref<2xf32>) // REDUCE_X_OP_ADD + REDUCE_ROW
+  tpp.reduce_add ins(%0) outs(%1) : (memref<4x2xf32>, memref<f32>) // REDUCE_X_OP_ADD + REDUCE_SCALAR
+  tpp.reduce_max ins(%0) outs(%1) : (memref<4x2xf32>, f32) // REDUCE_X_OP_MAX + REDUCE_SCALAR
+  tpp.reduce_max ins(%0) outs(%1) : (memref<2x4xf32>, memref<2x4xf32>) // Error, use COPY
+  tpp.reduce_max ins(%0) outs(%1) : (memref<2xf32>, memref<2x4xf32>) // Error, use BROADCAST
+```
+Should be fused with users.
+Should be safe to fuse with multiple users as they shouldn't alias output with an input of a different shape.
+It should be an error for `unary`/`binary`/`ternary` ops to alias with reduced operands.
+Can be lowered as `COPY` with `REDUCE` flags.
+
+## Transpose
+Transpose a shape into new memory.
+Type must have the same rank, 2, dims flipped.
+```mlir
+  tpp.transpose ins(%0) outs(%1) : memref<MxNxTy> -> memref<NxMxTy> // TRANSFORM_NORM_TO_NORMT
+```
+Shiould be fused with the user(s).
+GEMM ops have transposed versions, we should use this op to annotate operands.
+
+## BF16 Pack / Unpack
+Split fp32 into bf16 halfs.
```

Comment:
Unzip an fp32 tensor to two tensors with 16-bit entries in an elementwise fashion / Zip two tensors with 16-bit entries into an fp32 tensor in an elementwise fashion